File: framework/lib/panel.dart
part of uxml;

/**
* Panel provides a container with a visual state(maximized,minimized) and a
* title.
*
* @author ferhat@ (Ferhat Buyukkokten)
*/
class Panel extends ContentContainer {
  static ElementDef panelElementDef;
  /** Title property definition */
  static PropertyDefinition titleProperty;

  /** ViewState property definition */
  static PropertyDefinition viewStateProperty;

  /** MaximizeEnabled property definition */
  static PropertyDefinition maximizeEnabledProperty;

  /** MinimizeEnabled property definition */
  static PropertyDefinition minimizeEnabledProperty;

  /** MoveEnabled property definition */
  static PropertyDefinition moveEnabledProperty;

  /** Panel moved event definition */
  static EventDef movedEvent;

  // Windowing behaviour:
  // 1- A Panel can be displayed modally using the show method and
  //    closed using the close method.
  //
  // 2- For maximize, minimize, restore operations, the viewState change
  //    is handled by the element hosting the Panel or by the application
  //    itself (for modal top level panels).

  /** ViewState constant for normal view */
  static const int VIEW_STATE_NORMAL = 0;

  /** ViewState constant for maximized view */
  static const int VIEW_STATE_MAXIMIZED = 1;

  /** ViewState constant for maximized view */
  static const int VIEW_STATE_MINIMIZED = 2;

  /**
  * Standard framework id for maximize button.
  */
  static const String MAXIMIZE_BUTTON_ID = "maximizePanelButton";

  /**
   * Standard framework id for minimize button.
   */
  static const String MINIMIZE_BUTTON_ID = "minimizePanelButton";

  /**
  * Standard framework id for maximize button chrome.
  */
  static const String MAXIMIZE_CHROME_ID = "maximizePanelChrome";

  /**
  * Standard framework id for maximize button chrome.
  */
  static const String MINIMIZE_CHROME_ID = "minimizePanelChrome";


  /** Mouse position at time of drag_start */
  Coord _mouseDelta;
  EventArgs _movedEventArgs;

  bool _isMousePressed = false;

  /**
   * Constructor
   */
  Panel() : super() {
  }

  /**
  * Gets or returns title of panel.
  */
  String get title {
    return getProperty(titleProperty);
  }

  set title(String value) {
    setProperty(titleProperty, value);
  }

  /** Overrides UIElement.onMouseDown. */
  void onMouseDown(MouseEventArgs mouseArgs) {
    if (moveEnabled) {
      captureMouse();
      _mouseDelta = mouseArgs.getMousePosition(this);
      _isMousePressed = true;
      mouseArgs.handled = true;
    }
    if (parent is UIElementContainer) {
      UIElementContainer container = parent;
      container.bringToFront(this);
    }
  }

  /** Overrides UIElement.onMouseDown. */
  void onMouseMove(MouseEventArgs mouseArgs) {
    if (_isMousePressed) {
      Coord mousePosition = mouseArgs.getMousePosition(this.parent);
      Canvas.setChildLeft(this, mousePosition.x - _mouseDelta.x);
      Canvas.setChildTop(this, mousePosition.y - _mouseDelta.y);
    }
  }

  /** Overrides UIElement.onMouseDown. */
  void onMouseUp(MouseEventArgs mouseArgs) {
    if (_isMousePressed) {
      _isMousePressed = false;
      releaseMouse();
      if (_movedEventArgs == null) {
        _movedEventArgs = new EventArgs(this);
      }
      notifyListeners(movedEvent, _movedEventArgs);
    }
  }

  /**
   * Activated the panel in modal form.
   */
  void show() {
    Application.current._addModalPanel(this);
  }

  /**
   * Closes panel.
   */
  void close() {
    Application.current._removeModalPanel(this);
    viewState = VIEW_STATE_NORMAL;
    _movedEventArgs = null;
    super.close();
  }

  /**
   * Maximizes panel.
   */
  void maximize() {
    viewState = VIEW_STATE_MAXIMIZED;
  }

  /**
   * Minimizes panel.
   */
  void minimize() {
    viewState = VIEW_STATE_MINIMIZED;
  }

  /**
   * Restores panel.
   */
  void restore() {
    viewState = VIEW_STATE_NORMAL;
  }

  /**
   * Sets or returns view state of panel.
   */
  set viewState(int state) {
    setProperty(viewStateProperty, state);
  }

  int get viewState {
    return getProperty(viewStateProperty);
  }

  /** Gets or sets if panel maximize is enabled */
  bool get maximizeEnabled {
    return getProperty(maximizeEnabledProperty);
  }

  set maximizeEnabled(bool value) {
    setProperty(maximizeEnabledProperty, value);
  }

  /** Gets or sets if panel minimize is enabled */
  bool get minimizeEnabled {
    return getProperty(minimizeEnabledProperty);
  }

  set minimizeEnabled(bool value) {
    setProperty(minimizeEnabledProperty, value);
  }

  /** Gets or sets if panel move is enabled */
  bool get moveEnabled {
    return getProperty(moveEnabledProperty);
  }

  set moveEnabled(bool value) {
    setProperty(moveEnabledProperty, value);
  }

  static void _maxMinEnabledChanged(Object target,
                                    Object property,
                                    Object oldValue,
                                    Object newValue) {
    Panel panel = target;
    UIElement element = panel.getElement(MAXIMIZE_BUTTON_ID);
    if (element != null) {
      element.visible = panel.maximizeEnabled;
    }
    element = panel.getElement(MINIMIZE_BUTTON_ID);
    if (element != null) {
      element.visible = panel.minimizeEnabled;
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => panelElementDef;

  /** Registers component. */
  static void registerPanel() {
    titleProperty = ElementRegistry.registerProperty("Title",
        PropertyType.STRING, PropertyFlags.NONE, null, "");
    viewStateProperty = ElementRegistry.registerProperty("ViewState",
        PropertyType.INT, PropertyFlags.NONE, null, VIEW_STATE_NORMAL);
    maximizeEnabledProperty = ElementRegistry.registerProperty(
        "MaximizeEnabled", PropertyType.BOOL, PropertyFlags.NONE,
        _maxMinEnabledChanged, true);
    minimizeEnabledProperty = ElementRegistry.registerProperty(
        "MinimizeEnabled", PropertyType.BOOL, PropertyFlags.NONE,
        _maxMinEnabledChanged, true);
    moveEnabledProperty = ElementRegistry.registerProperty("moveEnabled",
        PropertyType.BOOL, PropertyFlags.NONE, null, false);
    movedEvent = new EventDef("Moved", Route.DIRECT);
    panelElementDef = ElementRegistry.register("Panel",
        ContentContainer.contentcontainerElementDef,
        [titleProperty, viewStateProperty, maximizeEnabledProperty,
        minimizeEnabledProperty, moveEnabledProperty], [movedEvent]);
    // Override default isFocusGroup property.
    UIElement.isFocusGroupProperty.overrideDefaultValue(panelElementDef,true);

  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Panel.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;
import com.hello.uxml.tools.framework.events.EventDefinition;
import com.hello.uxml.tools.framework.events.EventManager;
import com.hello.uxml.tools.framework.events.MouseEventArgs;

import java.util.EnumSet;

/**
 * Implements a container that provides windowing services.
 *
 * @author ferhat
 */
public class Panel extends ContentContainer {

  /** ViewState constant for normal view */
  public static final int VIEW_STATE_NORMAL = 0;

  /** ViewState constant for maximized view */
  public static final int VIEW_STATE_MAXIMIZED = 1;

  /** ViewState constant for maximized view */
  public static final int VIEW_STATE_MINIMIZED = 2;

  /**
   * Standard framework id for maximize button.
   */
   public static final String MAXIMIZE_BUTTON_ID = "maximizePanelButton";

   /**
    * Standard framework id for minimize button.
    */
   public static final String MINIMIZE_BUTTON_ID = "minimizePanelButton";

   /**
   * Standard framework id for maximize button chrome.
   */
   public static final String MAXIMIZE_CHROME_ID = "maximizePanelChrome";

   /**
   * Standard framework id for maximize button chrome.
   */
   public static final String MINIMIZE_CHROME_ID = "minimizePanelChrome";

  /** Title Property Definition */
  public static PropertyDefinition titlePropDef = PropertySystem.register("Title",
      String.class,
      Panel.class,
      new PropertyData("", EnumSet.of(PropertyFlags.None, PropertyFlags.Localizable)));

  /** ViewState Property Definition */
  public static PropertyDefinition viewStatePropDef = PropertySystem.register("ViewState",
      Integer.class,
      Panel.class,
      new PropertyData(VIEW_STATE_NORMAL, EnumSet.of(PropertyFlags.None)));

  /** MaximizeEnabled Property Definition */
  public static PropertyDefinition maximizeEnabledPropDef = PropertySystem.register(
      "MaximizeEnabled", Boolean.class, Panel.class,
      new PropertyData(true, EnumSet.of(PropertyFlags.None)));

  /** MinimizeEnabled Property Definition */
  public static PropertyDefinition minimizeEnabledPropDef = PropertySystem.register(
      "MinimizeEnabled", Boolean.class, Panel.class,
      new PropertyData(true, EnumSet.of(PropertyFlags.None)));

  /** MoveEnabled Property Definition */
  public static PropertyDefinition moveEnabledPropDef = PropertySystem.register(
      "MoveEnabled", Boolean.class, Panel.class,
      new PropertyData(false, EnumSet.of(PropertyFlags.None)));

  /** Panel moved event definition */
  public static EventDefinition movedEvent = EventManager.register(
      "Moved", Panel.class, EventArgs.class, Panel.class, null);

  private EventArgs movedEventArgs = null;
  private boolean isMousePressed = false;
  private Point mouseDelta;

  /** Sets or returns title of panel */
  public String getTitle() {
    return (String) getProperty(titlePropDef);
  }

  public void setTitle(String title) {
    setProperty(titlePropDef, title);
  }

  /** Sets or returns view state */
  public int getViewState() {
    return ((Integer) getProperty(viewStatePropDef)).intValue();
  }

  public void setViewState(int viewState) {
    setProperty(viewStatePropDef, viewState);
  }

  /**
   * Sets or returns whether panel can be maximized.
   */
  public void setMaximizeEnabled(boolean value) {
    setProperty(maximizeEnabledPropDef, value);
  }

  public boolean getMaximizeEnabled() {
     return ((Boolean) getProperty(maximizeEnabledPropDef)).booleanValue();
  }

  /**
   * Sets or return if panel can be moved with mouse.
   */
  public void setMoveEnabled(boolean value) {
    setProperty(moveEnabledPropDef, value);
  }

  public boolean getMoveEnabled() {
    return ((Boolean) getProperty(moveEnabledPropDef)).booleanValue();
  }

  /**
   * Sets or returns whether panel can be minimized.
   */
  public void setMinimizeEnabled(boolean value) {
    setProperty(maximizeEnabledPropDef, value);
  }

  public boolean getMinimizeEnabled() {
     return ((Boolean) getProperty(minimizeEnabledPropDef)).booleanValue();
  }

  /**
   * Restore panel view.
   */
  public void restore() {
    setViewState(VIEW_STATE_NORMAL);
  }

  /**
   * Maximizes panel view.
   */
  public void maximize() {
    setViewState(VIEW_STATE_MAXIMIZED);
  }

  /**
   * Minimizes panel view.
   */
  public void minimize() {
    setViewState(VIEW_STATE_MINIMIZED);
  }

  @Override
  protected void onMouseDown(MouseEventArgs e) {
    if (getMoveEnabled()) {
      captureMouse();
      mouseDelta = e.getMousePosition(this);
      isMousePressed = true;
      e.setHandled(true);
    }
  }

  @Override
  protected void onMouseMove(MouseEventArgs e) {
    if (isMousePressed) {
      Point mousePosition = e.getMousePosition((UIElement) this.parent);
      Canvas.setChildLeft(this, mousePosition.x - mouseDelta.x);
      Canvas.setChildTop(this, mousePosition.y - mouseDelta.y);
      if (movedEventArgs == null) {
        movedEventArgs = new EventArgs(this, movedEvent);
        notifyListeners(movedEvent, movedEventArgs);
      }
    }
  }
}
